PyInstaller/building/build_main.py

```diff
@@ -482,8 +515,19 @@ def assemble(self):
         self.pure._code_cache = self.graph.get_code_objects()
 
         # Add remaining binary dependencies - analyze Python C-extensions and what DLLs they depend on.
+        #
+        # Up until this point, we did very best not to import the packages into the main process. However, a package
+        # may set up additional library search paths during its import (e.g., by modifying PATH or calling the
+        # add_dll_directory() function on Windows, or modifying LD_LIBRARY_PATH on Linux). In order to reliably
+        # discover dynamic libraries, we therefore require an environment with all packages imported. We achieve that
+        # by gathering list of all collected packages, and spawn an isolated process, in which we first import all
+        # the packages from the list, and then peform search for dynamic libraries.
```

Comment:
Does this not put us back into the danger zone that is multiple Qt variants being loaded in the same process leading to the wrong DLLs being collected?

Reply:
It does. But if we collect multiple variants of Qt-based packages for some reason (e.g., `Qt5Agg` `matplotlib` backend in an environment where we have both PyQt5 and PySide2 installed), we are risking undefined behavior in other places as well: 
* only one set of shared libs will end up collected in `sys._MEIPASS`
* multiple rthooks are executed at run-time, so QT_PLUGIN_PATH and QML2_IMPORT_PATH are set to the value of  the last-executed qthook

I suppose the practical issues stemming from this are now somewhat mitigated by the fact that both packages are using Qt 5.15.2, so those shared libraries and plugins are (at least in theory) interchangeable.

Reply:
So I think this problem existed before this PR (because hooks ended up importing Qt packages during modulegraph analysis) and continues to exist in it (although admittedly, it might manifest itself differently, as the order of imports is very likely to change).

One way to solve it could be implementing some sort of mutual-exclusion mechanism for groups of packages: 
in a post-processing step (after modulegraph analysis but before binary dependency analysis), we'd go over collected packages, and if a higher-ranking package from a group is to be collected, it automatically suppresses other packages from that group.